File: LeetCode/572.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def isSubtree(self, root: Optional[TreeNode], subRoot: Optional[TreeNode]) -> bool:
        if not subRoot:
            return True
        if not root:
            return False
        if self.sameTree(root, subRoot):
            return True
        return self.isSubtree(root.left, subRoot) or self.isSubtree(root.right, subRoot)

    def sameTree(self, a, b):
        if not a and not b:
            return True
        if not a or not b:
            return False
        if not a.left and not b.left and not a.right and not b.right and a.val == b.val:
            return True
        return a.val == b.val and self.sameTree(a.left, b.left) and self.sameTree(a.right, b.right)

File: leetcode-new/572.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution
{
public:
    bool isSubtree(TreeNode *root, TreeNode *subRoot)
    {
        if (!subRoot)
            return true;
        if (!root)
            return false;
        if (helper(root, subRoot))
            return true;
        return isSubtree(root->left, subRoot) || isSubtree(root->right, subRoot);
    }

    bool helper(TreeNode *t1, TreeNode *t2)
    {
        if (!t1 && !t2)
            return true;
        if ((t1 && t2) && (t1->val == t2->val))
        {
            return helper(t1->left, t2->left) && helper(t1->right, t2->right);
        }
        return false;
    }
};
